File: js/bestSum_tab.js
// bestSum(targetSum, numbers)
//
// bestSum(8, [2, 3, 5]) -> [3, 5]

function bestSum(targetSum, numbers) {
	const table = Array(targetSum + 1).fill(null)
	table[0] = []

	for (let i=0; i <= targetSum; i++) {
		if (table[i])
			for (let num of numbers) {
				const minComb = [...table[i] , num]
				if (!table[i+num] || minComb.length < table[i+num].length)
					table[i + num] = minComb
			}
	}

	return table[targetSum]
}

console.log(bestSum(7, [4, 3]))
console.log(bestSum(15, [4, 3, 5]))
console.log(bestSum(10, [4, 3, 5]))
console.log(bestSum(12, [7, 5, 3, 4]))
console.log(bestSum(100, [1,2,5,25]))

File: cpp/bestSum_tab.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> bestSum(int target, const vector<int>& numbers) {
	vector<vector<int>> table (target + 1);
	table[0] = {0};

	for (int i = 0; i <= target; i++) {
		if (!table[i].empty()) {
			if (table[i][0] == 0) table[i].pop_back();
			for (auto num : numbers) {
				if (num + i <= target && (table[num + i].empty() || table[num + i].size() > table[i].size())) {
					table[num + i] = table[i];
					table[num + i].push_back(num);
				}
			}
		}
	}

	return table[target];
}

vector<int> bestSum(int target, const vector<int>&& numbers) {
	return bestSum(target, numbers);
}

int main() {
	// auto result = bestSum(7, {4, 3, 7});
	auto result = bestSum(8, {5, 3, 7, 1, 2});
	// auto result = bestSum(300, {7, 14});
	for (auto r : result)
		cout << r << " ";
	cout << endl;
}
